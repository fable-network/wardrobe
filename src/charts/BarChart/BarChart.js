import React from 'react';
import PropTypes from 'prop-types';
import styled, { css, withTheme } from 'styled-components';
import { theme as defaultTheme } from '../../theme';
import HighChart from '../HighChart';

const highlightedCss = css`
  .highcharts-point.highcharts-color-0 {
    fill: ${p => p.theme.grey05};

    &.highcharts-point-hover {
      fill: ${p => p.theme.primaryActive};
      fill-opacity: 1;
    }
  }
`;

const Wrapper = styled.div`
  .ft-wardrobe-bar-chart {
    .highcharts-point.highcharts-color-0 {
      &,
      &.highcharts-point-hover {
        transition: fill 250ms;
        stroke-width: 0;
        fill: ${p => p.theme.primaryActive};
        fill-opacity: 1;
      }
    }
    ${p => p.highlighted && highlightedCss};

    .highcharts-yaxis-grid {
      .highcharts-grid-line {
        stroke-dasharray: 4;
      }
    }
    .highcharts-xaxis .highcharts-axis-line {
      stroke: ${p => p.theme.grey04};
      stroke-width: 2px;
    }
    div.highcharts-xaxis-labels {
      display: flex;
      font-size: ${p => p.theme.fontSizeSmall};
      line-height: ${p => p.theme.lineHeightSmall};
      font-weight: 400;
      color: ${p => p.theme.grey01};
      text-align: center;

      .ft-label-block {
        display: block;

        .ft-label.ft-label-category {
          display: block;
          max-width: 100%;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
          text-align: right;
        }

        &.ft-label-block-disabled {
          .ft-label.ft-label-category {
            color: ${p => p.theme.grey04};
          }
        }
      }
    }

    .highcharts-tooltip {
      stroke: ${p => p.theme.grey04};

      .highcharts-tooltip-box {
        fill: ${p => p.theme.white};
        fill-opacity: 1;
      }

      .ft-tooltip {
        /* This is a hack to avoid labels showing *through* tooltips */
        display: block;
        position: relative;
        padding: 16px;
        top: -7px;
        left: -7px;
        margin-bottom: -17px;
        margin-right: -17px;
        background-color: ${p => p.theme.white};

        .ft-tooltip-header {
          display: block;
          font-size: ${p => p.theme.fontSizeSmall};
          line-height: ${p => p.theme.lineHeightSmall};
          font-weight: 500;
        }

        .ft-tooltip-body {
          font-size: ${p => p.theme.fontSizeSmall};
          line-height: ${p => p.theme.lineHeightSmall};
          font-weight: 400;
        }
      }
    }

    .highcharts-credits {
      display: none;
    }
  }
`;

class BarChart extends React.PureComponent {
  state = { highlightIndex: null };

  handleMouseOver = ({ target }) => {
    const { index } = target;
    this.setState({ highlightIndex: index });
  };

  handleMouseOut = () => {
    this.setState({ highlightIndex: null });
  };

  renderLabel = (pos, category) => {
    const { highlightIndex } = this.state;
    const className = `ft-label-block ${
      highlightIndex !== null && highlightIndex !== pos ? 'ft-label-block-disabled' : ''
    }`;
    return `
      <span class="${className}" data-index="${pos}" title="${category}">
        <span class="ft-label ft-label-category">
          ${category}
        </span>
      </span>
    `;
  };

  renderValue = (value, total) => {
    const percent = ((100 * value) / total).toFixed(0);
    return `${percent}%`;
  };

  render() {
    const { renderLabel, renderValue, handleMouseOver, handleMouseOut } = this;
    const { data, title, tooltip, orientation, allowUpdate, theme = defaultTheme } = this.props;
    if (!data) {
      return null;
    }
    const { highlightIndex } = this.state;
    const pointWidth = data.length > 10 && orientation === 'horizontal' ? 6 : 8;
    const total = data.reduce((sum, item) => sum + item.y, 0);
    const options = {
      title: title && {
        text: title,
      },
      chart: {
        type: orientation === 'horizontal' ? 'bar' : 'column',
        backgroundColor: theme.white,
        styledMode: true,
        groupPadding: 0,
      },
      plotOptions: {
        series: {
          groupPadding: 0,
          events: {
            mouseOver: handleMouseOver,
            mouseOut: handleMouseOut,
          },
        },
        bar: {
          borderRadius: pointWidth / 2,
          pointWidth,
          stickyTracking: true,
        },
        column: {
          borderRadius: pointWidth / 2,
          pointWidth,
          stickyTracking: true,
        },
      },
      xAxis: {
        // It requires categories in advance to plan for space allocation
        categories: data.map(point => point.name),
        gridLineWidth: 0,
        tickWidth: 0,
        offset: 1,
        lineWidth: 0,
        labels: {
          useHTML: true,
          formatter: function formatter() {
            return renderLabel(this.pos, this.value);
          },
        },
      },
      yAxis: {
        title: false,
        labels: {
          formatter: function formatter() {
            return renderValue(this.value, total);
          },
        },
        gridLineWidth: 0,
      },
      legend: {
        enabled: false,
      },
      tooltip: {
        enabled: true,
        outside: false, // Don't change unless you want to re-do the whole tooltip styling
        useHTML: true,
        shadow: true,
        backgroundColor: theme.white,
        borderColor: theme.grey04,
        borderRadius: 0,
        padding: 8,
        formatter: function formatter() {
          const percentage = (100 * this.y) / total;
          const body = tooltip({ ...this, percentage });
          return `<span class="ft-tooltip">
            <span class="ft-tooltip-header">${this.x}</span>
            <span class="ft-tooltip-body">${body}</span>
          </span>`;
        },
      },
      series: [{ data }],
    };
    return (
      <Wrapper highlighted={highlightIndex !== null}>
        <HighChart
          options={options}
          className="ft-wardrobe-bar-chart"
          allowChartUpdate={allowUpdate}
        />
      </Wrapper>
    );
  }
}

BarChart.defaultProps = {
  title: false,
  orientation: 'horizontal',
  tooltip: ({ y }) => y,
  allowUpdate: false,
};

BarChart.propTypes = {
  orientation: PropTypes.oneOf(['horizontal', 'vertical']),
  title: PropTypes.string,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      y: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
  tooltip: PropTypes.func,
  /**
   * Allows HighCharts chart update. Use it if you plan to provide new data.
   */
  allowUpdate: PropTypes.bool,
  // from withTheme
  theme: PropTypes.object.isRequired,
};

export default withTheme(BarChart);
