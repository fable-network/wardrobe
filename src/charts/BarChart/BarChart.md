A bar chart for many data points. Example usage: category mix, price points.

```jsx
<BarChart title="Category mix" data={require('../fixtures/bar-chart-category-mix.js').default} />
```

Even more data points and a custom tooltip format:

```jsx
<BarChart
  title="Category mix"
  data={require('../fixtures/bar-chart-category-mix-many.js').default}
  tooltip={({ y, percentage }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
/>
```

Vertical bars (e.g. for Price points):

```jsx
<BarChart
  orientation="vertical"
  title="Price point"
  data={require('../fixtures/bar-chart-price-point.js').default}
  tooltip={({ y, percentage }) =>
    `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`
  }
/>
```
