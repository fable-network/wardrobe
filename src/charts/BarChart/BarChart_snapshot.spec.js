/* eslint-disable */
// This file was automatically generated
import React from 'react';
import 'jest-styled-components';
import BarChart from './BarChart';


jest.mock('../../charts/HighChart', () => () => <high-charts />);

describe('BarChart Component', () => {
  describe('snapshots', () => {
    it('should match snapshot', () => {
    
      const wrapper1 = mount(
        <BarChart title="Category mix" data={require('../fixtures/bar-chart-category-mix.js').default} />
      );
      expect(wrapper1).toMatchSnapshot();

      const wrapper2 = mount(
        <BarChart
          title="Category mix"
          data={require('../fixtures/bar-chart-category-mix-many.js').default}
          tooltip={({ y, percentage }) => `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`}
        />
      );
      expect(wrapper2).toMatchSnapshot();

      const wrapper3 = mount(
        <BarChart
          orientation="vertical"
          title="Price point"
          data={require('../fixtures/bar-chart-price-point.js').default}
          tooltip={({ y, percentage }) =>
            `Quantity: ${y}<br>Percent: ${percentage.toFixed(0)}%`
          }
        />
      );
      expect(wrapper3).toMatchSnapshot();

    });
  });
});
