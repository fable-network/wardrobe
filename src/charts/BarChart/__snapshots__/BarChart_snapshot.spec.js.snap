// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart Component snapshots should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(BarChart)
    data={
      Array [
        Object {
          "name": "Jeans",
          "y": 2600,
        },
        Object {
          "name": "Tops",
          "y": 1900,
        },
        Object {
          "name": "Shirts",
          "y": 1625,
        },
        Object {
          "name": "Cardigans",
          "y": 1475,
        },
        Object {
          "name": "Pull-overs",
          "y": 1000,
        },
        Object {
          "name": "Jackets and tops",
          "y": 800,
        },
        Object {
          "name": "Chino pants",
          "y": 400,
        },
        Object {
          "name": "Ridiculous White Board Fancy Irregular Shorts",
          "y": 200,
        },
      ]
    }
    title="Category mix"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "data": Array [
        Object {
          "name": "Jeans",
          "y": 2600,
        },
        Object {
          "name": "Tops",
          "y": 1900,
        },
        Object {
          "name": "Shirts",
          "y": 1625,
        },
        Object {
          "name": "Cardigans",
          "y": 1475,
        },
        Object {
          "name": "Pull-overs",
          "y": 1000,
        },
        Object {
          "name": "Jackets and tops",
          "y": 800,
        },
        Object {
          "name": "Chino pants",
          "y": 400,
        },
        Object {
          "name": "Ridiculous White Board Fancy Irregular Shorts",
          "y": 200,
        },
      ],
      "orientation": "horizontal",
      "theme": undefined,
      "title": "Category mix",
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "data": Array [
          Object {
            "name": "Jeans",
            "y": 2600,
          },
          Object {
            "name": "Tops",
            "y": 1900,
          },
          Object {
            "name": "Shirts",
            "y": 1625,
          },
          Object {
            "name": "Cardigans",
            "y": 1475,
          },
          Object {
            "name": "Pull-overs",
            "y": 1000,
          },
          Object {
            "name": "Jackets and tops",
            "y": 800,
          },
          Object {
            "name": "Chino pants",
            "y": 400,
          },
          Object {
            "name": "Ridiculous White Board Fancy Irregular Shorts",
            "y": 200,
          },
        ],
        "orientation": "horizontal",
        "theme": undefined,
        "title": "Category mix",
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`BarChart Component snapshots should match snapshot 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(BarChart)
    data={
      Array [
        Object {
          "name": "Jeans",
          "y": 2600,
        },
        Object {
          "name": "Tops",
          "y": 1900,
        },
        Object {
          "name": "Shirts",
          "y": 1625,
        },
        Object {
          "name": "Cardigans",
          "y": 1475,
        },
        Object {
          "name": "Pull-overs",
          "y": 1000,
        },
        Object {
          "name": "Jackets and tops",
          "y": 800,
        },
        Object {
          "name": "Chino pants",
          "y": 400,
        },
        Object {
          "name": "Ridiculous White Board Fancy Irregular Shorts",
          "y": 200,
        },
        Object {
          "name": "Hats",
          "y": 100,
        },
        Object {
          "name": "Panamas",
          "y": 90,
        },
        Object {
          "name": "Accessories",
          "y": 80,
        },
        Object {
          "name": "Jewelry",
          "y": 50,
        },
      ]
    }
    title="Category mix"
    tooltip={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "data": Array [
        Object {
          "name": "Jeans",
          "y": 2600,
        },
        Object {
          "name": "Tops",
          "y": 1900,
        },
        Object {
          "name": "Shirts",
          "y": 1625,
        },
        Object {
          "name": "Cardigans",
          "y": 1475,
        },
        Object {
          "name": "Pull-overs",
          "y": 1000,
        },
        Object {
          "name": "Jackets and tops",
          "y": 800,
        },
        Object {
          "name": "Chino pants",
          "y": 400,
        },
        Object {
          "name": "Ridiculous White Board Fancy Irregular Shorts",
          "y": 200,
        },
        Object {
          "name": "Hats",
          "y": 100,
        },
        Object {
          "name": "Panamas",
          "y": 90,
        },
        Object {
          "name": "Accessories",
          "y": 80,
        },
        Object {
          "name": "Jewelry",
          "y": 50,
        },
      ],
      "orientation": "horizontal",
      "theme": undefined,
      "title": "Category mix",
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "data": Array [
          Object {
            "name": "Jeans",
            "y": 2600,
          },
          Object {
            "name": "Tops",
            "y": 1900,
          },
          Object {
            "name": "Shirts",
            "y": 1625,
          },
          Object {
            "name": "Cardigans",
            "y": 1475,
          },
          Object {
            "name": "Pull-overs",
            "y": 1000,
          },
          Object {
            "name": "Jackets and tops",
            "y": 800,
          },
          Object {
            "name": "Chino pants",
            "y": 400,
          },
          Object {
            "name": "Ridiculous White Board Fancy Irregular Shorts",
            "y": 200,
          },
          Object {
            "name": "Hats",
            "y": 100,
          },
          Object {
            "name": "Panamas",
            "y": 90,
          },
          Object {
            "name": "Accessories",
            "y": 80,
          },
          Object {
            "name": "Jewelry",
            "y": 50,
          },
        ],
        "orientation": "horizontal",
        "theme": undefined,
        "title": "Category mix",
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`BarChart Component snapshots should match snapshot 3`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithTheme(BarChart)
    data={
      Array [
        Object {
          "name": "€10-",
          "y": 50,
        },
        Object {
          "name": "€10—19.99",
          "y": 90,
        },
        Object {
          "name": "€20—29.99",
          "y": 70,
        },
        Object {
          "name": "€30—39.99",
          "y": 30,
        },
        Object {
          "name": "€40—49.99",
          "y": 104,
        },
        Object {
          "name": "€50—59.99",
          "y": 60,
        },
        Object {
          "name": "€60—69.99",
          "y": 200,
        },
        Object {
          "name": "€70—79.99",
          "y": 176,
        },
        Object {
          "name": "€80—89.99",
          "y": 43,
        },
        Object {
          "name": "€90—99.99",
          "y": 182,
        },
        Object {
          "name": "€100—109.99",
          "y": 90,
        },
        Object {
          "name": "€110—119.99",
          "y": 21,
        },
        Object {
          "name": "€120—129.99",
          "y": 37,
        },
        Object {
          "name": "€130—139.99",
          "y": 23,
        },
        Object {
          "name": "€140—149.99",
          "y": 34,
        },
        Object {
          "name": "€150+",
          "y": 10,
        },
      ]
    }
    orientation="vertical"
    title="Price point"
    tooltip={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "allowUpdate": false,
      "data": Array [
        Object {
          "name": "€10-",
          "y": 50,
        },
        Object {
          "name": "€10—19.99",
          "y": 90,
        },
        Object {
          "name": "€20—29.99",
          "y": 70,
        },
        Object {
          "name": "€30—39.99",
          "y": 30,
        },
        Object {
          "name": "€40—49.99",
          "y": 104,
        },
        Object {
          "name": "€50—59.99",
          "y": 60,
        },
        Object {
          "name": "€60—69.99",
          "y": 200,
        },
        Object {
          "name": "€70—79.99",
          "y": 176,
        },
        Object {
          "name": "€80—89.99",
          "y": 43,
        },
        Object {
          "name": "€90—99.99",
          "y": 182,
        },
        Object {
          "name": "€100—109.99",
          "y": 90,
        },
        Object {
          "name": "€110—119.99",
          "y": 21,
        },
        Object {
          "name": "€120—129.99",
          "y": 37,
        },
        Object {
          "name": "€130—139.99",
          "y": 23,
        },
        Object {
          "name": "€140—149.99",
          "y": 34,
        },
        Object {
          "name": "€150+",
          "y": 10,
        },
      ],
      "orientation": "vertical",
      "theme": undefined,
      "title": "Price point",
      "tooltip": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "allowUpdate": false,
        "data": Array [
          Object {
            "name": "€10-",
            "y": 50,
          },
          Object {
            "name": "€10—19.99",
            "y": 90,
          },
          Object {
            "name": "€20—29.99",
            "y": 70,
          },
          Object {
            "name": "€30—39.99",
            "y": 30,
          },
          Object {
            "name": "€40—49.99",
            "y": 104,
          },
          Object {
            "name": "€50—59.99",
            "y": 60,
          },
          Object {
            "name": "€60—69.99",
            "y": 200,
          },
          Object {
            "name": "€70—79.99",
            "y": 176,
          },
          Object {
            "name": "€80—89.99",
            "y": 43,
          },
          Object {
            "name": "€90—99.99",
            "y": 182,
          },
          Object {
            "name": "€100—109.99",
            "y": 90,
          },
          Object {
            "name": "€110—119.99",
            "y": 21,
          },
          Object {
            "name": "€120—129.99",
            "y": 37,
          },
          Object {
            "name": "€130—139.99",
            "y": 23,
          },
          Object {
            "name": "€140—149.99",
            "y": 34,
          },
          Object {
            "name": "€150+",
            "y": 10,
          },
        ],
        "orientation": "vertical",
        "theme": undefined,
        "title": "Price point",
        "tooltip": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
